### PR TITLE
Vdc storage profile limit update

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Ref: #461, #497, #404
Dep: https://github.com/vmware/go-vcloud-director/pull/340

* change allows to update VDC storage profile. Before VDC storage profile was used only for create.
* new computed field `storage_used_mb`
* fix documentation for #488

### CODE REVIEW

File: vcd/datasource_vcd_org_vdc.go [hunk 1]
```diff
@@ -118,6 +118,11 @@ func datasourceVcdOrgVdc() *schema.Resource {
 							Computed:    true,
 							Description: "True if this is default storage profile for this VDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified.",
 						},
+						"storage_used_mb": {
```
Comment (Didainius) [inline vcd/datasource_vcd_org_vdc.go:121]: In other places we used `*in_mb`. Maybe would be good to be consistent:
`size_in_mb`
`limit_in_mb`

Also description would have something similar:
"The size of the disk in MB."

File: vcd/resource_vcd_org_vdc.go [hunk 2]
```diff
@@ -110,10 +110,11 @@ func resourceVcdOrgVdc() *schema.Resource {
 				Description: "True if this VDC is enabled for use by the organization VDCs. Default is true.",
 			},
 			"storage_profile": &schema.Schema{
-				Type:     schema.TypeList,
-				Required: true,
-				ForceNew: true,
-				MinItems: 1,
+				Type:        schema.TypeList,
```
Comment (Didainius) [inline vcd/resource_vcd_org_vdc.go:113]: Just a thought here as I see it now. Is this not going to cause problems if/when storage profiles are not returned in the same order during READ?

File: vcd/resource_vcd_org_vdc.go [hunk 3]
```diff
@@ -137,9 +138,13 @@ func resourceVcdOrgVdc() *schema.Resource {
 							Required:    true,
 							Description: "True if this is default storage profile for this VDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified.",
 						},
+						"storage_used_mb": {
```
Comment (Didainius) [inline vcd/resource_vcd_org_vdc.go:141]: Same comments as in datasource

**Review by Didainius — CHANGES_REQUESTED** [review]
I think I am missing update test for resource.

**Comment by vbauzysvmware** [comment]
> I think I am missing update test for resource.

it is in `org_vdc_common_test.go`

File: vcd/datasource_vcd_org_vdc.go (hunk 1 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_org_vdc.go:121]: good point, changed.

File: vcd/resource_vcd_org_vdc.go (hunk 2 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:113]: no complains until now, yet exists such possibility

File: vcd/resource_vcd_org_vdc.go (hunk 3 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go]: changed

File: vcd/resource_vcd_org_vdc.go (hunk 2 above)
Comment (dataclouder) [reply vcd/resource_vcd_org_vdc.go:114]: There is a [complaint](https://github.com/vmware/terraform-provider-vcd/issues/404)

Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go]: converted to typeSet

File: vcd/resource_vcd_org_vdc.go [hunk 4]
```diff
@@ -970,8 +1011,8 @@ func getVcdVdcInput(d *schema.ResourceData, vcdClient *VCDClient) (*types.VdcCon
 	}
 	computeCapacity := computeCapacityList[0].(map[string]interface{})
 
-	vdcStorageProfilesConfigurations := d.Get("storage_profile").([]interface{})
-	if len(vdcStorageProfilesConfigurations) == 0 {
+	vdcStorageProfilesConfigurations := d.Get("storage_profile").(*schema.Set)
```
Comment (lvirbalas) [inline vcd/resource_vcd_org_vdc.go:1014]: Can this return nil?

File: vcd/resource_vcd_org_vdc.go [hunk 5]
```diff
@@ -111,10 +111,11 @@ func resourceVcdOrgVdc() *schema.Resource {
 				Description: "True if this VDC is enabled for use by the organization VDCs. Default is true.",
 			},
 			"storage_profile": &schema.Schema{
-				Type:     schema.TypeList,
-				Required: true,
-				ForceNew: true,
-				MinItems: 1,
+				Type:        schema.TypeSet,
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:114]: to fix this #404

File: vcd/resource_vcd_org_vdc.go [hunk 6]
```diff
@@ -463,6 +468,7 @@ func getComputeStorageProfiles(vcdClient *VCDClient, profile *types.VdcStoragePr
 		if vdcStorageProfileDetails.ProviderVdcStorageProfile != nil {
 			storageProfileData["name"] = vdcStorageProfileDetails.ProviderVdcStorageProfile.Name
 		}
+		storageProfileData["storage_used_in_mb"] = vdcStorageProfileDetails.StorageUsedMB
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:471]: well, I will need to change in govcd this

[the remark this answers is not in the document]
Comment (lvirbalas) [reply vcd/resource_vcd_org_vdc.go:471]: That's why it's better not to hurry merging govcd PR until Terraform PR is done.

[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:471]: `MB` hit 18 and `Mb` hit 3 cases in types.go. Not sure about renaming anymore.

File: vcd/datasource_vcd_org_vdc.go [hunk 7]
```diff
@@ -113,6 +113,11 @@ func datasourceVcdOrgVdc() *schema.Resource {
 							Computed:    true,
 							Description: "True if this is default storage profile for this VDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified.",
 						},
+						"storage_used_in_mb": {
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "Storage used, in Megabytes",
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/datasource_vcd_org_vdc.go:119]: changed

File: vcd/resource_vcd_org_vdc.go [hunk 8]
```diff
@@ -138,9 +139,13 @@ func resourceVcdOrgVdc() *schema.Resource {
 							Required:    true,
 							Description: "True if this is default storage profile for this VDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified.",
 						},
+						"storage_used_in_mb": {
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "Storage used, in Megabytes",
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:145]: changed

File: vcd/resource_vcd_org_vdc.go (hunk 4 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_org_vdc.go:1014]: if key doesn't have value, then default empty valued returned. Nil is returned of not existing key requested

Comment (Didainius) [reply vcd/resource_vcd_org_vdc.go:1014]: As far as I see `storage_profile` is mandatory so it should be safe

**Review by Didainius — APPROVED** [review]
Thanks. Vdc tests passed.

**Review by lvirbalas — APPROVED** [review]

**Review by dataclouder — APPROVED** [review]

**Pull request merged**